### PR TITLE
add query logging and other logging changes (close #507, #2171)

### LINKED ISSUE #507: Verbose logging
via @praveenweb:

"we should enable verbose logging mode (via flag/env), where instead of just the query_hash, we get the actual query for better debugging".

### LINKED ISSUE #570: test migration path across different server versions
The migration path is currently untested by the build system

### PR DESCRIPTION
**Pull request by ecthiender**
### Description

Following changes are introduced in this PR:
1. enables logging of the GraphQL query and the generated SQL
2. adds a new flag to enable/disable log-types (`--enabled-log-types`)
3. logs all of the server configuration during startup
4. changes existing log structures (http and websocket) (**breaking change**)
5. adds a new flag to set logging levels
6. changes existing `http-log` to log level `error` in case of errors (previously was `info`) (*breaking change*)

### Query logging
A new log-type `query-log` is introduced which logs the GraphQL query and the generated SQL.
This can be enabled via the new introduced flag `--enabled-log-types`. (More details below)

Along with existing `http-log`, this will log the following (as a separate log type `query-log`, on every GraphQL request over HTTP and websocket):

1. The entire GraphQL query with variables
2. Generated SQL statements (only for queries, not for mutations/subscriptions or remote schema queries)
3. The operation name (if provided in the GraphQL request)

Generates a request id (if `x-request-id` header is not found in request headers), and logs it in both `http-log` and `query-log`, in case of HTTP transport. Similarly, in both `websocket-log` and `query-log`, in case of websocket transport. This helps to co-relate requests between http/websocket logs and the query logs.

### New flag `--enabled-log-types` 
A new  flag `--enabled-log-types` (and env var `HASURA_GRAPHQL_ENABLED_LOG_TYPES`) is introduced to enable/disable specific log types.

For example:
 `--enabled-log-types='query-log, http-log, websocket-log'`. (defaults to `startup, http-log, websocket-log, webhook-log`).

All the log-types are documented here: https://deploy-preview-1835--hasura-docs.netlify.com/graphql/manual/deployment/logging.html#different-log-types

### Logs server config on startup
Logs all server configuration during startup (previously only a subset of all the config was logged), and also logs the startup time in an object instead of string (#2171)

### New flag `--log-level`
New flag `--log-level` (and env var `HASURA_GRAPHQL_LOG_LEVEL`) is introduced to set desired logging level by the user. 

For example:
`--log-level error`. (default: `info`.)

**NOTE**: The above will only enable error logs to be printed.  So even if the user has enabled `query-log` it won't be printed as the level of `query-log`s are `info`. Only errors will be printed in `http-log` or `websocket-log`s.

**breaking change**: previously errors in `http-log` were printed at the `info` log level, changed that to `error`.

More details are documented: https://deploy-preview-1835--hasura-docs.netlify.com/graphql/manual/deployment/logging.html#logging-levels

### Changes existing log structure (breaking)
This issue also introduces the following changes:

- Changes the `http-log` structure to make it more easy to analyze (more details below)
- Rename `ws-handler` to `websocket-log` and changes in `websocket-log` structure (details below)
- Removes `query_hash` from logs

~~**NOTE**: Please note that only the structure has changed, and this has nothing to do with verbose logging. Verbose logging only enables the `query-log` to get printed, nothing of `http-log` or `websocket-log` is affected by verbose logging.~~

### Migration path

Previously, there were two main kinds of logs for every request - `http-log` and `ws-handler` for HTTP and websockets respectively. (The other logs being, logs during startup, event-trigger logs, schema-sync logs, jwk-refresh logs etc.). Now on enabling verbose log, `query-log` also gets printed for each request (over http and websocket). This also changes the structure of `http-log` and `query-log`.

A summary of the changes in a table format can be viewed here: https://deploy-preview-1835--hasura-docs.netlify.com/graphql/manual/deployment/logging.html#summary-of-the-changes

The migration path is documented: https://deploy-preview-1835--hasura-docs.netlify.com/graphql/manual/deployment/logging.html#migration-path-of-logs-from-v1-0-0-beta-2-to-newer

<details> 
<summary> Details of log strucures </summary>
#### The structure of the `http-log` has changed

Older, on success :

```json
{
  "timestamp": "2019-06-07T12:04:16.713+0000",
  "level": "info",
  "type": "http-log",
  "detail": {
    "status": 200,
    "query_hash": "e9006e6750ebaa77da775ae4fc60227d3101b03e",
    "http_version": "HTTP/1.1",
    "query_execution_time": 0.408548571,
    "request_id": "1ad0c61b-1431-410e-818e-99b57822bd2b",
    "url": "/v1/graphql",
    "ip": "106.51.72.39",
    "response_size": 204,
    "user": {
      "x-hasura-role": "admin"
    },
    "method": "POST",
    "detail": null
  }
}
```

Newer, on success:

```json
{
  "timestamp": "2019-05-30T23:40:24.654+0530",
  "level": "info",
  "type": "http-log",
  "detail": {
    "operation": {
      "query_execution_time": 0.009240042,
      "user_vars": {
        "x-hasura-role": "user"
      },
      "request_id": "072b3617-6653-4fd5-b5ee-580e9d098c3d",
      "response_size": 105,
      "error": null,
      "query": null
    },
    "http_info": {
      "status": 200,
      "http_version": "HTTP/1.1",
      "url": "/v1/graphql",
      "ip": "127.0.0.1",
      "method": "POST"
    }
  }
}
```

Older on error:

```json
{
  "timestamp": "2019-06-07T12:24:05.166+0000",
  "level": "info",
  "type": "http-log",
  "detail": {
    "status": 200,
    "query_hash": "511894cc797a2b5cef1c84f106a038ea7bc8436d",
    "http_version": "HTTP/1.1",
    "query_execution_time": 2.34687e-4,
    "request_id": "02d695c7-8a2d-4a45-84dd-8b61b7255807",
    "url": "/v1/graphql",
    "ip": "106.51.72.39",
    "response_size": 138,
    "user": {
      "x-hasura-role": "admin"
    },
    "method": "POST",
    "detail": {
      "error": {
        "path": "$.selectionSet.todo.selectionSet.completedx",
        "error": "field \"completedx\" not found in type: 'todo'",
        "code": "validation-failed"
      },
      "request": "{\"query\":\"query {\\n  todo {\\n    id\\n    title\\n    completedx\\n  }\\n}\",\"variables\":null}"
    }
  }
}
```

Newer, on error:

```json
{
  "timestamp": "2019-05-29T15:22:37.834+0530",
  "level": "info",
  "type": "http-log",
  "detail": {
    "operation": {
      "query_execution_time": 0.000656144,
      "user_vars": {
        "x-hasura-role": "user",
        "x-hasura-user-id": "1"
      },
      "error": {
        "path": "$.selectionSet.profile.selectionSet.usernamex",
        "error": "field \"usernamex\" not found in type: 'profile'",
        "code": "validation-failed"
      },
      "request_id": "072b3617-6653-4fd5-b5ee-580e9d098c3d",
      "response_size": 142,
      "query": {
        "variables": {
          "limit": 10
        },
        "operationName": "getProfile",
        "query": "query getProfile($limit: Int!) { profile(limit: $limit, where:{username: {_like: \"%a%\"}}) { usernamex} }"
      }
    },
    "http_info": {
      "status": 200,
      "http_version": "HTTP/1.1",
      "url": "/v1/graphql",
      "ip": "127.0.0.1",
      "method": "POST"
    }
  }
}
```

#### The structure for `ws-handler` has changed, and `ws-handler` has been renamed to `websocket-log`

Older, on success:

```json
{
  "timestamp": "2019-06-07T12:35:40.652+0000",
  "level": "info",
  "type": "ws-handler",
  "detail": {
    "event": {
      "type": "operation",
      "detail": [
        "1",
        null,
        {
          "type": "started"
        }
      ]
    },
    "websocket_id": "11dea559-6554-4598-969a-00b48545950f",
    "jwt_expiry": null,
    "msg": null,
    "user": {
      "x-hasura-role": "admin"
    }
  }
}
```

Newer, on success:

```json
{
  "timestamp": "2019-06-10T10:52:54.247+0530",
  "level": "info",
  "type": "websocket-log",
  "detail": {
    "event": {
      "type": "operation",
      "detail": {
        "request_id": "d2ede87d-5cb7-44b6-8736-1d898117722a",
        "operation_id": "1",
        "query": {
          "variables": {
            
          },
          "query": "subscription {\n  author {\n    name\n  }\n}\n"
        },
        "operation_type": {
          "type": "started"
        },
        "operation_name": null
      }
    },
    "connection_info": {
      "websocket_id": "f590dd18-75db-4602-8693-8150239df7f7",
      "jwt_expiry": null,
      "msg": null
    },
    "user_vars": {
      "x-hasura-role": "admin"
    }
  }
}
```

Older, when operation stops:

```json
{
  "timestamp": "2019-06-10T05:30:41.432+0000",
  "level": "info",
  "type": "ws-handler",
  "detail": {
    "event": {
      "type": "operation",
      "detail": [
        "1",
        null,
        {
          "type": "stopped"
        }
      ]
    },
    "websocket_id": "3f5721ee-1bc6-424c-841f-8ff8a326d9ef",
    "jwt_expiry": null,
    "msg": null,
    "user": {
      "x-hasura-role": "admin"
    }
  }
}
```

Newer, when operations stops:

```json
{
  "timestamp": "2019-06-10T11:01:40.939+0530",
  "level": "info",
  "type": "websocket-log",
  "detail": {
    "event": {
      "type": "operation",
      "detail": {
        "request_id": null,
        "operation_id": "1",
        "query": null,
        "operation_type": {
          "type": "stopped"
        },
        "operation_name": null
      }
    },
    "connection_info": {
      "websocket_id": "7f782190-fd58-4305-a83f-8e17177b204e",
      "jwt_expiry": null,
      "msg": null
    },
    "user_vars": {
      "x-hasura-role": "admin"
    }
  }
}
```

Older, on error:

```json
{
  "timestamp": "2019-06-07T12:38:07.188+0000",
  "level": "info",
  "type": "ws-handler",
  "detail": {
    "event": {
      "type": "operation",
      "detail": [
        "1",
        null,
        {
          "type": "query_err",
          "detail": {
            "path": "$.selectionSet.todo.selectionSet.titlex",
            "error": "field \"titlex\" not found in type: 'todo'",
            "code": "validation-failed"
          }
        }
      ]
    },
    "websocket_id": "77558d9b-99f8-4c6a-b105-a5b08c96543b",
    "jwt_expiry": null,
    "msg": null,
    "user": {
      "x-hasura-role": "admin"
    }
  }
}
```

Newer, on error:

```json
{
  "timestamp": "2019-06-10T10:55:20.650+0530",
  "level": "info",
  "type": "websocket-log",
  "detail": {
    "event": {
      "type": "operation",
      "detail": {
        "request_id": "150e3e6a-e1a7-46ba-a9d4-da6b192a4005",
        "operation_id": "1",
        "query": {
          "variables": {
            
          },
          "query": "subscription {\n  author {\n    namex\n  }\n}\n"
        },
        "operation_type": {
          "type": "query_err",
          "detail": {
            "path": "$.selectionSet.author.selectionSet.namex",
            "error": "field \"namex\" not found in type: 'author'",
            "code": "validation-failed"
          }
        },
        "operation_name": null
      }
    },
    "connection_info": {
      "websocket_id": "49932ddf-e54d-42c6-bffb-8a57a1c6dcbe",
      "jwt_expiry": null,
      "msg": null
    },
    "user_vars": {
      "x-hasura-role": "admin"
    }
  }
}
```

The `query-log`:

```json
    {
      "timestamp": "2019-06-03T13:25:10.915+0530",
      "level": "info",
      "type": "query-log",
      "detail": {
        "request_id": "840f952d-c489-4d21-a87a-cc23ad17926a",
        "query": {
          "variables": {
            "limit": 10
          },
          "operationName": "getProfile",
          "query": "query getProfile($limit: Int!) {\n  profile(limit: $limit, where: {username: {_like: \"%a%\"}}) {\n    username\n  }\n  myusername: profile (where: {username: {_eq: \"foobar\"}}) {\n    username\n  }\n}\n"
        },
        "generated_sql": {
          "profile": {
            "prepared_arguments": ["{\"x-hasura-role\":\"admin\"}", "%a%"],
            "query": "SELECT  coalesce(json_agg(\"root\" ), '[]' ) AS \"root\" FROM  (SELECT  row_to_json((SELECT  \"_1_e\"  FROM  (SELECT  \"_0_root.base\".\"username\" AS \"username\"       ) AS \"_1_e\"      ) ) AS \"root\" FROM  (SELECT  *  FROM \"public\".\"profile\"  WHERE ((\"public\".\"profile\".\"username\") LIKE ($2))     ) AS \"_0_root.base\"     LIMIT 10 ) AS \"_2_root\"      "
          },
          "myusername": {
            "prepared_arguments": ["{\"x-hasura-role\":\"admin\"}", "foobar"],
            "query": "SELECT  coalesce(json_agg(\"root\" ), '[]' ) AS \"root\" FROM  (SELECT  row_to_json((SELECT  \"_1_e\"  FROM  (SELECT  \"_0_root.base\".\"username\" AS \"username\"       ) AS \"_1_e\"      ) ) AS \"root\" FROM  (SELECT  *  FROM \"public\".\"profile\"  WHERE ((\"public\".\"profile\".\"username\") = ($2))     ) AS \"_0_root.base\"      ) AS \"_2_root\"      "
          }
        }
      }
    }
```
</details>

More details about the log structures is in the docs.

### Affected components 
- Server
- Docs

### Related Issues
- #507
- #2171 

#### Query logging
Introduced a new log type called `query-log`. This is because the SQL generation happens in deeper modules and it is converted to a transaction that can be run. Changed those functions to return a transaction and the generated SQL. And instead of returning the generated SQL back up to the HTTP/websocket handler, we log it at the execution level, as a `query-log`.

The HTTP/websocket handler generates (if required) a request id (and logs it with its access log), which is then passed to the execution modules, which logs it with the query and SQL.

No SQL for mutations and subscriptions, doing this for mutations is hard, and the generated SQL is not readable anyway.

#### log-types
Added a new sum type `EngineLogType` to enumerate all the possible log-types currently in the system. Changed `EngineLog` to use this type instead of text.

* `alwaysOnLogTypes` is defined as a HashSet, which are the internal log types and can't be enabled/disabled by user config
* `defaultEnabledLogTypes` is defined as a HashSet, in terms of `alwaysOnLogTypes`.
* changes `LoggerCtx` to add the set of enabled log types. This is configured during the init process
* changes `mkLogger` function to get the enabled log types from `LoggerCtx` and then decide to log it or not
* other changes to `Hasura.Server.Init` and `Main` to parse the flag and fallback to default, initialize the `LoggerCtx` etc.

#### Query logging changes
Start with:
* `Hasura.GraphQL.Execute.Query` - most of the changes for getting the generated SQL is here
  * new type `PreparedSql` is introduced to capture the generated SQL query and prepared args
  * new type `ResolvedResp` to distinguish between SQL or bytestring response (incase of introspection)
  * new type alias `GeneratedSql` to have a map of the GraphQL field/alias and its corresponding `PreparedSql`
  * `mkCurPlanTx` function is changed to compute the `ResolvedResp` first and then return `LazyRespTx` and `GeneratedSql`
  * similarly to above `queryOpFromPlan` is changed
* New module: `Hasura.GraphQL.Logging`: contains graphql query logging related types and functions
* In module: `Hasura.GraphQL.Execute`, `ExecOp` type is changed to have the `GeneratedSql` for queries.
* In the various execution modules, the generated SQL and the query is logged from `ExecOp`.
  * new type `ExecutionCtx` is introduced to capture all the information needed by execution functions. Previously these were passed as arguments from the HTTP/Websocket handler. Now they are in a single ADT, and is passed via `ReaderT`.
* `Hasura.Server.App` module is changed significantly to pass the parsed query into `mkSpockAction`. Now `mkSpockAction` takes a `APIHandler` which can be a GET or POST. `mkSpockAction` parsed the request bytestring and then passed it to appropriate handler. This way `mkSpockAction` has access to the parsed query (to log incase of errors).

**Important modules to look at**:
* `Hasura.GraphQL.Execute.Query`
* `Hasura.GraphQL.Execute`
* `Hasura.GraphQL.Logging`
* `Hasura.GraphQL.Transport.HTTP`
* `Hasura.GraphQL.Transport.WebSocket`
* `Hasura.Server.App`
* `Hasura.Server.Logging`

#### Server configuration logging changes
* Changes to `serveOptsToLog` function to log everything in `ServeOptions`.
* Added appropriate JSON instances to relevant types.
* **Question**: is just a `show` on `TxIsolation` and `HostPreference` fine? Their output is: `"transaction_isolation":"ISOLATION LEVEL READ COMMITTED"`, `"server_host":"HostAny"` or `"server_host":"Host \"127.0.0.1\""`

Modules to look at:
* `Hasura.Server.Init`

#### `--enabled-log-type` changes
Start with:
* `Hasura.Logging` - introduces a new sum type for all possible log-types in the system. Changes existing `EngineLog` type to use this type instead of text. (Consequently, `ToEngineLog` are also changed)
  * `alwaysOnLogTypes` is defined as a HashSet, which are the internal log types and can't be enabled/disabled by user config
  * `defaultEnabledLogTypes` is defined as a HashSet, in terms of `alwaysOnLogTypes`.
  * changes `LoggerCtx` to add the set of enabled log types. This is configured during the init process
  * changes `mkLogger` function to get the enabled log types from `LoggerCtx` and then decide to log it or not
* other changes to `Hasura.Server.Init` and `Main` to parse the flag and fallback to default, initialize the `LoggerCtx` etc.

#### Log breaking changes
Changes to the strucuture of `http-log` and `ws-handler` (renamed to `websocket-log`) are in the modules:
* `Hasura.Server.Logging`
* `Hasura.GraphQL.Transport.Websocket`

Details to changes to strucutre is explained in details above and in docs.

#### Log level changes
Straightforward. Most of the code was already there. Changed `main` function in `Main` to pass in the log-level than hard-coding it in `defaultLoggerSettings` (in `Hasura.Logging`). Added flag and env var to take user config for log level.

**Important Modules to look at**:
* `Hasura.Server.Init`
* `Main`
* `Hasura.Logging`

### Steps to test and verify
1. Start graphql-engine with `--enable-log-types='startup,http-log,webhook-log,websocket-log,query-log'` flag.
2. Add tables and make queries over HTTP or Websocket.
3. Observe that the GraphQL query and generated SQL statements are logged.

### Limitations, known bugs & workarounds
Ideally, the `http-log` and `websocket-log` structure should have been the same. But due to the nature of HTTP and websocket requests and the transport protocol for GraphQL being different, their log structure also is different.

### CODE REVIEW

**Comment by ecthiender** [comment]
Few issues with current logging:

1. Right now logging is inconsistent among `http-log` and `ws-handler` logs. Ideally, they both should have same structure so that analyzing them is easier and straightforward. 
2. Starting with `/v1/graphql` all HTTP status are 200. Earlier strategy was to look at HTTP status codes to determine errors/failures. With `/v1/graphql` the approach will be convoluted - look for presence of `errors` key in nested JSON. We can probably do better by having a separate type of logs (say `http-error-log`).

File: server/src-lib/Hasura/Server/App.hs [hunk 1]
```diff
@@ -179,62 +188,83 @@ buildQCtx = do
...
 
   -- log result
-  logResult (Just userInfo) req reqBody logger (apiRespToLBS <$> modResult) $
-    Just (t1, t2)
+  logResult logger verboseLog (Just userInfo) requestId req (Just q)
+    (apiRespToLBS <$> modResult) $ Just (t1, t2)
+-- =======
+--   logResult (Just userInfo) req reqBody logger (apiRespToLBS <$> modResult) $
+--     Just (t1, t2)
+-- >>>>>>> ba96697abc5e22f2ab10d30b43062ea683618f9c
```
Comment (rakeshkky) [inline server/src-lib/Hasura/Server/App.hs]: merge conflict comments?

File: server/src-lib/Hasura/Server/Test.hs [hunk 2]
```diff
@@ -0,0 +1,33 @@
+module Hasura.Server.Test where
```
Comment (rakeshkky) [inline server/src-lib/Hasura/Server/Test.hs]: Remove this!

**Comment by rakeshkky** [comment]
split `mkSpockAction` to Handle `get` and `post` requests.

File: server/src-lib/Hasura/GraphQL/Logging.hs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  when (L.unVerboseLogging verbose) . liftIO . L.unLogger logger
+
+mkQueryLog :: RequestId -> GQLReqUnparsed -> Maybe EQ.GeneratedSql
+           -> QueryLog
+mkQueryLog reqId req sql = QueryLog (Just req) (EQ.encodeSql <$> sql) reqId
+
+data QueryLog
+  = QueryLog
+  { _qlQuery        :: !(Maybe GQLReqUnparsed)
+  , _qlGeneratedSql :: !(Maybe J.Value)
```
Comment (0x777) [inline server/src-lib/Hasura/GraphQL/Logging.hs]: J.Value -> GeneratedSql

File: server/src-lib/Hasura/GraphQL/Execute/Query.hs [hunk 4]
```diff
@@ -225,20 +231,64 @@ convertQuerySelSet varDefs fields = do
...
+        return $ RRSql res
+    return (alias, fldResp)
+
+  return (mkLazyRespTx resolved, mkGeneratedSql resolved)
+
+
+-- TODO: give better names
+data ResolvedResp
+  = RRRaw !B.ByteString
+  | RRSql !(Q.Query, [Q.PrepArg])
```
Comment (0x777) [inline server/src-lib/Hasura/GraphQL/Execute/Query.hs]: PreparedSql -> (Q.Query, Q.PrepArg)

File: server/src-lib/Hasura/GraphQL/Execute/Query.hs [hunk 5]
```diff
@@ -225,20 +231,64 @@ convertQuerySelSet varDefs fields = do
...
+
+mkGeneratedSql :: [(G.Alias, ResolvedResp)] -> GeneratedSql
+mkGeneratedSql resolved =
+  flip map resolved $ \(alias, node) ->
+    let res = case node of
+                RRRaw _         -> Nothing
+                RRSql (q, args) -> Just (q, args)
+    in (alias, res)
+
+encodeSql :: GeneratedSql -> J.Value
```
Comment (0x777) [inline server/src-lib/Hasura/GraphQL/Execute/Query.hs]: This has to go to `Logging` module.

File: server/src-lib/Hasura/GraphQL/Execute/LiveQuery.hs [hunk 6]
```diff
@@ -58,6 +58,12 @@ data LQOpts
   , _loFallbackOpts :: LQF.FallbackOpts
   } deriving (Show, Eq)
```
Comment (0x777) [inline server/src-lib/Hasura/GraphQL/Execute/LiveQuery.hs]: document that this is needed for server configuration log line

**Comment by ecthiender** [comment]
For changing the `--enable-verbose-log` to `--enable-log-types='query-log, http-log, websocket-log'`:

1. Each log-type also has log-levels like error, warn, info, debug.
2. Some log types (internal ones) like jwk-refresh-log, schema-sync log, event-trigger logs etc. are always on by default. Cannot be turned off or on by the `--enable-log-types` flag.
3. This information (from `--enable-log-types` flag) is part of `ServerCtx` after parsing.
4. Each log-type corresponds to each logging module (optional this for separation).
5. Each logging module exposes a function to log.
6. This function is used at different call-sites to log (with a log-level like debug, info, warn, error etc.).
7. This log function also takes in `ServerCtx`, and checks enabled log types and current log-level and decides if it should log or not.

**Comment by sryabkov** [comment]
Has any performance testing being done on this PR? Do you have any numbers to be able to predict the performance impact of keeping `query-log` always ON? Currently, hasura does log the query for failed requests, but not for successful ones. Instead it logs `null` 
Here is an example from [documentation](https://docs.hasura.io/1.0/graphql/manual/deployment/kubernetes/logging.html):

```
{"timestamp":"2018-10-09T11:20:32.054+0000", "level":"info", "type":"http-log", "detail":{"status":200, "query_hash":"01640c6dd131826cff44308111ed40d7fbd1cbed", "http_version":"HTTP/1.1", "query_execution_time":3.0177627e-2, "request_id":null, "url":"/v1/graphql", "user":{"x-hasura-role":"admin"}, "ip":"127.0.0.1", "response_size":209329, "method":"POST", "detail":null}}
```

> the last field, `detail` is null

File: server/src-lib/Hasura/GraphQL/Transport/HTTP.hs [hunk 7]
```diff
@@ -2,53 +2,58 @@ module Hasura.GraphQL.Transport.HTTP
...
+  E.ExecutionCtx logger _ pgExecCtx _ _ _ _ _ <- ask
   respE <- liftIO $ runExceptT $ case resolvedOp of
-    E.ExOpQuery tx    ->
+    E.ExOpQuery tx genSql  -> do
+      -- log the generated SQL and the graphql query
+      liftIO $ logGraphqlQuery logger $ QueryLog query genSql reqId
       runLazyTx' pgExecCtx tx
-    E.ExOpMutation tx ->
+    E.ExOpMutation tx -> do
+      -- log the generated SQL and the graphql query
```
Comment (rakeshkky) [inline server/src-lib/Hasura/GraphQL/Transport/HTTP.hs]: Only graphql query right?

File: server/src-lib/Hasura/Server/Logging.hs [hunk 8]
```diff
@@ -1,45 +1,44 @@
 -- This is taken from wai-logger and customised for our use
+{-# LANGUAGE OverloadedStrings #-}
```
Comment (rakeshkky) [inline server/src-lib/Hasura/Server/Logging.hs]: We can remove this extension.

File: server/src-lib/Hasura/Server/Logging.hs [hunk 9]
```diff
@@ -118,118 +117,133 @@ instance ToJSON WebHookLog where
...
-  , alHttpVersion  = httpVersion req
-  , alDetail       = mDetail
-  , alRequestId    = bsToTxt <$> getRequestId req
-  , alHasuraUser   = userVars <$> userInfoM
-  , alResponseSize = size
-  , alResponseTime = realToFrac <$> diffTime
-  , alQueryHash    = queryHash
-  }
+  -> HttpLog
+mkAccessLog userInfoM reqId req res mTimeT =
```
Comment (rakeshkky) [inline server/src-lib/Hasura/Server/Logging.hs]: `mkHttpLog`?

File: server/src-lib/Hasura/Server/Logging.hs [hunk 10]
```diff
@@ -118,118 +117,133 @@ instance ToJSON WebHookLog where
...
-           , "query_hash" .= qh
-           , "response_size" .= rs
-           , "query_execution_time" .= rt
            ]
 
-data LogDetail
-  = LogDetail
-  { _ldQuery :: !TBS.TByteString
-  , _ldError :: !Value
+-- | Information about a GraphQL operation over HTTP
```
Comment (rakeshkky) [inline server/src-lib/Hasura/Server/Logging.hs]: Can `OperationLog` is used for `/v1/query`?

If yes then this comment is not correct.

File: server/src-lib/Hasura/Server/Logging.hs (hunk 9 above)
Comment (rakeshkky) [reply server/src-lib/Hasura/Server/Logging.hs]: `mkAccessLog` and `mkErrorLog` both gives `HttpLog`. One is on `success` and the other is on `error`. I think it's better to define functions accordingly
Like:-
- `mkSuccessHttpLog`
- `mkErrorHttpLog`

File: server/src-lib/Hasura/Server/Logging.hs [hunk 11]
```diff
@@ -118,118 +117,133 @@ instance ToJSON WebHookLog where
...
+           , olQueryExecutionTime = respTime
+           , olQuery = toJSON <$> query
+           , olError = Just $ toJSON err
+           }
+  in HttpLog L.LevelError $ HttpAccessLog http op
+  where
+    status = qeStatus err
+    (respTime, respSize) = genLogInfo (encode err) mTimeT
+
+genLogInfo
```
Comment (rakeshkky) [inline server/src-lib/Hasura/Server/Logging.hs]: It's actually generating query exec time and response size.

Comment (rakeshkky) [reply server/src-lib/Hasura/Server/Logging.hs]: Split into two functions

File: server/src-lib/Hasura/Logging.hs [hunk 12]
```diff
@@ -25,20 +28,66 @@ import qualified Data.Aeson.Casing     as J
...
+  ]
+
+-- the default enabled log-types
+defaultEnabledLogTypes :: Set.HashSet EngineLogType
+defaultEnabledLogTypes =
+  Set.union alwaysOnLogTypes $
+  Set.fromList [ELTStartup, ELTHttpLog, ELTWebhookLog, ELTWebsocketLog]
+
+-- log types that can be set by the user
+userAllowedLogTypes :: [String]
```
Comment (rakeshkky) [inline server/src-lib/Hasura/Logging.hs]: `[EngineLogType]`

File: server/src-lib/Hasura/GraphQL/Transport/WebSocket.hs [hunk 13]
```diff
@@ -239,71 +273,78 @@ onStart serverEnv wsConn (StartMsg opId q) msgRaw = catchAndIgnore $ do
...
 
     WSConnData userInfoR opMap errRespTy = WS.getData wsConn
 
-    logOpEv opDet =
-      logWSEvent logger wsConn $ EOperation opId (_grOperationName q) opDet
+    logOpEv opTy reqId =
+      logWSEvent logger wsConn $ EOperation opDet
+      where
+        opDet = OperationDetails opId reqId (_grOperationName q) opTy gq
+        gq = case opTy of
```
Comment (rakeshkky) [inline server/src-lib/Hasura/GraphQL/Transport/WebSocket.hs]: better naming?

**Comment by BenHizak** [comment]
What is the minimum Hasura version that will include this feature?

**Comment by sryabkov** [comment]
@BenHizak Judging by the [release notes](https://github.com/hasura/graphql-engine/releases/tag/v1.0.0-beta.3), this feature has been released in `v1.0.0-beta.3`

**Pull request merged**